.github/workflows/check_license_header.yaml

```diff
@@ -0,0 +1,85 @@
+name: Check license
+
+on: [pull_request]
+
+jobs:
+  check-licenses:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/javascript-action@v1
+      - uses: actions/checkout@v2
+
+      - name: Check file extensions
+        if: always()          # needed so that it runs even if previous one fails
+        run: |
+          # highlight if in these folders there are files with other extensions
+          find src include test miniapp -type f                                   \
+               ! '('                                                              \
+                       -name '*.cpp'                                              \
+                   -o  -name '*.h'                                                \
+                   -o  -name '*.h.in'                                             \
+                   -o  -name '*.tpp'                                              \
+                   -o  -name '*.cu'                                               \
+                   -o  -name 'CMakeLists.txt'                                     \
+               ')'                                                                \
+            > result-extension.check
+          # Generate an error message for each file with an unkown extension
+          for filepath in `cat result-extension.check`; do                        \
+            echo "::error file=$filepath,line=1::check extension of $filepath";   \
+          done
+
+          test ! -s result-extension.check
+
+      - name: Check source code
+        if: always()          # needed so that it runs even if previous one fails
+        run: |
+          # Compare first lines of each source file with reference in misc/HEADER
+          find src include test miniapp -type f                                   \
+               '('                                                                \
+                       -name '*.cpp'                                              \
+                   -o  -name '*.h'                                                \
+                   -o  -name '*.h.in'                                             \
+                   -o  -name '*.tpp'                                              \
+                   -o  -name '*.cu'                                               \
+               ')'                                                                \
+            | xargs -I{} sh -c                                                    \
+                ".github/check_license.sh misc/HEADER {} > /dev/null || echo {}"  \
+            > result-cpp.check
+
+          # Generate an error message for each offending file
+          for filepath in `cat result-cpp.check`; do                              \
+            echo "::error file=$filepath,line=1::check the license in $filepath"; \
+          done
+
+          test ! -s result-cpp.check
```

Comment:
LGTM!

If you would like to avoid the temporary files `<sth>.check` and to avoid spawning a child-process when using `cat`, perhaps you can do something like:

```
<long_find_and_xargs_command> | \
while read filepath; do
    echo $filepath
done
```

Also some refactoring (into a helper script) on the error message and the find-xargs combo can be made if you would like to shorten the code a bit, but this is a non-blocker and can be done in a future PR.

Reply:
That's a good point, but it is just for the cpp-like check because there is just one "contributor" to `result-cpp.check`, right? Because, for the #-based check the contributors are cmake and python files.

Do you have any suggestion also for them?

Reply:
There's another point. After generating error messages within the `for`, there is also to return an error code based on the same input we pipeline to the `for`.

I don't know if it fits somehow in your proposal.